File: rad/engine/02_catalog/model/dependencies.go
package model

import "slices"

// CatalogDependencies is the immutable compatibility contract carried by a
// bound physical plan. Names are diagnostic only; physical IDs and expected
// fence generations determine identity and admission.
//
// This is internal execution metadata. It is neither persisted nor exposed
// through PIR or LIR, so it deliberately has no serialization tags.
type CatalogDependencies struct {
	TableExistence []TableExistenceDependency
	ColumnValues   []ColumnValueDependency
	IndexAccess    []IndexAccessDependency
	WriteProtocols []WriteProtocolDependency
}

// TableExistenceDependency pins the existence fence of one physical table.
type TableExistenceDependency struct {
	TableID    string
	TableName  string
	Generation uint64
}

// ColumnValueDependency pins one physical column's decoding and missing-value
// semantics.
type ColumnValueDependency struct {
	TableID    string
	TableName  string
	ColumnID   string
	ColumnName string
	Generation uint64
}

// IndexAccessDependency pins the lifetime and representation of one physical
// index selected by a plan.
type IndexAccessDependency struct {
	TableID    string
	TableName  string
	IndexID    string
	IndexName  string
	Generation uint64
}

// WriteProtocolDependency pins the complete mutation obligations for one
// table.
type WriteProtocolDependency struct {
	TableID    string
	TableName  string
	Generation uint64
}

// Empty reports whether the manifest admits no catalog compatibility fences.
func (d CatalogDependencies) Empty() bool {
	return len(d.TableExistence) == 0 &&
		len(d.ColumnValues) == 0 &&
		len(d.IndexAccess) == 0 &&
		len(d.WriteProtocols) == 0
}

// Clone returns an independently owned copy of the dependency manifest.
func (d CatalogDependencies) Clone() CatalogDependencies {
	return CatalogDependencies{
		TableExistence: slices.Clone(d.TableExistence),
		ColumnValues:   slices.Clone(d.ColumnValues),
		IndexAccess:    slices.Clone(d.IndexAccess),
		WriteProtocols: slices.Clone(d.WriteProtocols),
	}
}

// AddTableRead records the value semantics needed to decode columns from one
// live physical table. An empty columns slice is valid for row-counting plans:
// they depend on the table and its rows, but decode no cells.
func (d *CatalogDependencies) AddTableRead(table Table, columns []Column) {
	d.addTableExistence(TableExistenceDependency{
		TableID: table.ID, TableName: table.Name, Generation: table.ExistenceGeneration,
	})
	for _, column := range columns {
		d.addColumnValue(ColumnValueDependency{
			TableID: table.ID, TableName: table.Name,
			ColumnID: column.ID, ColumnName: column.Name, Generation: column.ValueGeneration,
		})
	}
}

// AddIndexRead extends a table read with the physical index access path whose
// lifetime the plan relies upon.
func (d *CatalogDependencies) AddIndexRead(table Table, index Index, columns []Column) {
	d.AddTableRead(table, columns)
	d.addIndexAccess(IndexAccessDependency{
		TableID: table.ID, TableName: table.Name,
		IndexID: index.ID, IndexName: index.Name, Generation: index.AccessGeneration,
	})
}

// AddTableWrite records the complete current row definition and immutable
// write protocol a mutation must honor. Ready-index maintenance and active
// schema-transition obligations are members of that protocol.
func (d *CatalogDependencies) AddTableWrite(table Table) {
	d.AddTableRead(table, table.Columns)
	d.addWriteProtocol(WriteProtocolDependency{
		TableID: table.ID, TableName: table.Name, Generation: table.WriteProtocolGeneration,
	})
}

// Merge adds every dependency from other, deduplicating identical fences.
func (d *CatalogDependencies) Merge(other CatalogDependencies) {
	for _, dependency := range other.TableExistence {
		d.addTableExistence(dependency)
	}
	for _, dependency := range other.ColumnValues {
		d.addColumnValue(dependency)
	}
	for _, dependency := range other.IndexAccess {
		d.addIndexAccess(dependency)
	}
	for _, dependency := range other.WriteProtocols {
		d.addWriteProtocol(dependency)
	}
}

func (d *CatalogDependencies) addTableExistence(dependency TableExistenceDependency) {
	if slices.ContainsFunc(d.TableExistence, func(existing TableExistenceDependency) bool {
		return existing.TableID == dependency.TableID && existing.Generation == dependency.Generation
	}) {
		return
	}
	d.TableExistence = append(d.TableExistence, dependency)
}

func (d *CatalogDependencies) addColumnValue(dependency ColumnValueDependency) {
	if slices.ContainsFunc(d.ColumnValues, func(existing ColumnValueDependency) bool {
		return existing.TableID == dependency.TableID &&
			existing.ColumnID == dependency.ColumnID &&
			existing.Generation == dependency.Generation
	}) {
		return
	}
	d.ColumnValues = append(d.ColumnValues, dependency)
}

func (d *CatalogDependencies) addIndexAccess(dependency IndexAccessDependency) {
	if slices.ContainsFunc(d.IndexAccess, func(existing IndexAccessDependency) bool {
		return existing.TableID == dependency.TableID &&
			existing.IndexID == dependency.IndexID &&
			existing.Generation == dependency.Generation
	}) {
		return
	}
	d.IndexAccess = append(d.IndexAccess, dependency)
}

func (d *CatalogDependencies) addWriteProtocol(dependency WriteProtocolDependency) {
	if slices.ContainsFunc(d.WriteProtocols, func(existing WriteProtocolDependency) bool {
		return existing.TableID == dependency.TableID && existing.Generation == dependency.Generation
	}) {
		return
	}
	d.WriteProtocols = append(d.WriteProtocols, dependency)
}

File: src/engine/02_catalog/model/dependencies.rs
use super::{Column, Index, Table};
use crate::engine::catalog::identity::{
    AccessGeneration, ColumnId, ExistenceGeneration, IndexId, TableId, ValueGeneration,
    WriteProtocolGeneration,
};

/// Immutable compatibility fences carried by a bound physical plan. Names are
/// diagnostic only; physical identities and generations determine admission.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct CatalogDependencies {
    pub table_existence: Vec<TableExistenceDependency>,
    pub column_values: Vec<ColumnValueDependency>,
    pub index_access: Vec<IndexAccessDependency>,
    pub write_protocols: Vec<WriteProtocolDependency>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct TableExistenceDependency {
    pub table_id: TableId,
    pub table_name: String,
    pub generation: ExistenceGeneration,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ColumnValueDependency {
    pub table_id: TableId,
    pub table_name: String,
    pub column_id: ColumnId,
    pub column_name: String,
    pub generation: ValueGeneration,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct IndexAccessDependency {
    pub table_id: TableId,
    pub table_name: String,
    pub index_id: IndexId,
    pub index_name: String,
    pub generation: AccessGeneration,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct WriteProtocolDependency {
    pub table_id: TableId,
    pub table_name: String,
    pub generation: WriteProtocolGeneration,
}

impl CatalogDependencies {
    pub fn is_empty(&self) -> bool {
        self.table_existence.is_empty()
            && self.column_values.is_empty()
            && self.index_access.is_empty()
            && self.write_protocols.is_empty()
    }

    pub fn add_table_read(&mut self, table: &Table, columns: &[Column]) {
        push_unique(
            &mut self.table_existence,
            TableExistenceDependency {
                table_id: table.id.clone(),
                table_name: table.name.clone(),
                generation: table.existence_generation,
            },
            |left, right| left.table_id == right.table_id && left.generation == right.generation,
        );
        for column in columns {
            push_unique(
                &mut self.column_values,
                ColumnValueDependency {
                    table_id: table.id.clone(),
                    table_name: table.name.clone(),
                    column_id: column.id.clone(),
                    column_name: column.name.clone(),
                    generation: column.value_generation,
                },
                |left, right| {
                    left.table_id == right.table_id
                        && left.column_id == right.column_id
                        && left.generation == right.generation
                },
            );
        }
    }

    pub fn add_index_read(&mut self, table: &Table, index: &Index, columns: &[Column]) {
        self.add_table_read(table, columns);
        push_unique(
            &mut self.index_access,
            IndexAccessDependency {
                table_id: table.id.clone(),
                table_name: table.name.clone(),
                index_id: index.id.clone(),
                index_name: index.name.clone(),
                generation: index.access_generation,
            },
            |left, right| {
                left.table_id == right.table_id
                    && left.index_id == right.index_id
                    && left.generation == right.generation
            },
        );
    }

    pub fn add_table_write(&mut self, table: &Table) {
        self.add_table_read(table, &table.columns);
        push_unique(
            &mut self.write_protocols,
            WriteProtocolDependency {
                table_id: table.id.clone(),
                table_name: table.name.clone(),
                generation: table.write_protocol_generation,
            },
            |left, right| left.table_id == right.table_id && left.generation == right.generation,
        );
    }

    pub fn merge(&mut self, other: &Self) {
        for dependency in &other.table_existence {
            push_unique(
                &mut self.table_existence,
                dependency.clone(),
                |left, right| {
                    left.table_id == right.table_id && left.generation == right.generation
                },
            );
        }
        for dependency in &other.column_values {
            push_unique(
                &mut self.column_values,
                dependency.clone(),
                |left, right| {
                    left.table_id == right.table_id
                        && left.column_id == right.column_id
                        && left.generation == right.generation
                },
            );
        }
        for dependency in &other.index_access {
            push_unique(&mut self.index_access, dependency.clone(), |left, right| {
                left.table_id == right.table_id
                    && left.index_id == right.index_id
                    && left.generation == right.generation
            });
        }
        for dependency in &other.write_protocols {
            push_unique(
                &mut self.write_protocols,
                dependency.clone(),
                |left, right| {
                    left.table_id == right.table_id && left.generation == right.generation
                },
            );
        }
    }
}

fn push_unique<T>(values: &mut Vec<T>, value: T, same: impl Fn(&T, &T) -> bool) {
    if !values.iter().any(|existing| same(existing, &value)) {
        values.push(value);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::engine::catalog::identity::{DefinitionGeneration, LogicalIndexId, SchemaId};
    use crate::engine::catalog::model::{ScalarType, Table};

    fn fixture() -> (Table, Index) {
        let column = Column {
            id: "c1".into(),
            schema_id: SchemaId::new(1).unwrap(),
            name: "id".into(),
            value_generation: 4.into(),
            scalar_type: ScalarType::Text,
            nullable: false,
            format: String::new(),
            insert_default: None,
            missing_value: None,
        };
        let index = Index {
            id: "i1".into(),
            logical_id: LogicalIndexId::from("li1"),
            definition_generation: DefinitionGeneration::from(3),
            access_generation: AccessGeneration::from(5),
            name: "primary".into(),
            columns: vec!["id".into()],
            column_ids: vec!["c1".into()],
            unique: true,
            ..Index::default()
        };
        let table = Table {
            id: "t1".into(),
            schema_id: SchemaId::new(1).unwrap(),
            name: "users".into(),
            definition_generation: 2.into(),
            existence_generation: 6.into(),
            write_protocol_generation: 7.into(),
            columns: vec![column],
            primary_key: vec!["id".into()],
            indexes: vec![index.clone()],
            foreign_keys: Vec::new(),
            constraints: Vec::new(),
        };
        (table, index)
    }

    #[test]
    fn table_write_records_all_required_fence_classes_once() {
        let (table, _) = fixture();
        let mut dependencies = CatalogDependencies::default();
        dependencies.add_table_write(&table);
        dependencies.add_table_write(&table);
        assert_eq!(dependencies.table_existence.len(), 1);
        assert_eq!(dependencies.column_values.len(), 1);
        assert_eq!(dependencies.write_protocols.len(), 1);
    }

    #[test]
    fn index_read_adds_access_to_the_table_and_value_fences() {
        let (table, index) = fixture();
        let mut dependencies = CatalogDependencies::default();
        dependencies.add_index_read(&table, &index, &table.columns);
        assert_eq!(dependencies.table_existence.len(), 1);
        assert_eq!(dependencies.column_values.len(), 1);
        assert_eq!(dependencies.index_access.len(), 1);
    }
}
